feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/utils/integrityCheck/IntegrityCheckSession.kt

```diff
@@ -0,0 +1,121 @@
+package io.novafoundation.nova.feature_dapp_impl.utils.integrityCheck
+
+import io.novafoundation.nova.common.data.network.NetworkApiCreator
+import io.novafoundation.nova.common.data.storage.Preferences
+import io.novafoundation.nova.common.utils.IntegrityService
+import io.novafoundation.nova.common.utils.ensureSuffix
+import io.novafoundation.nova.common.utils.sha256
+import io.novafoundation.nova.common.utils.toBase64
+import io.novafoundation.nova.feature_dapp_impl.utils.integrityCheck.IntegrityCheckSession.Callback
+import java.util.UUID
+
+class IntegrityCheckSessionFactory(
+    private val apiCreator: NetworkApiCreator,
+    private val preferences: Preferences,
+    private val integrityService: IntegrityService
+) {
+
+    fun createSession(
+        baseUrl: String,
+        callback: Callback
+    ) = IntegrityCheckSession(
+        baseUrl,
+        apiCreator.create(IntegrityCheckApi::class.java, baseUrl.ensureSuffix("/")),
+        preferences,
+        integrityService,
+        callback
+    )
+}
+
+private const val PREFS_APP_INTEGRITY_ID = "PREFS_APP_INTEGRITY_ID"
+private const val PREFS_ATTESTATION_PASSED = "PREFS_ATTESTATION_PASSED"
+
+class IntegrityCheckSession(
+    private val baseUrl: String,
+    private val integrityCheckApi: IntegrityCheckApi,
+    private val preferences: Preferences,
+    private val integrityService: IntegrityService,
+    private var callback: Callback?
+) {
+
+    interface Callback {
+        fun sendVerificationRequest(challenge: String, appIntegrityId: String, signature: String)
+    }
+
+    suspend fun startIntegrityCheck() {
+        if (!isAttestationPassed()) {
+            runAttestation()
+        }
+
+        runVerifying()
+    }
+
+    suspend fun restartIntegrityCheck() {
+        runAttestation()
+
+        runVerifying()
+    }
+
+    fun removeCallback() {
+        callback = null
+    }
```

Comment:
This looks strange - why do we nullify somthing that has been passed in the constructor? As I see we only use it together with nullifying session itself - i think this solely is enough to enable gc to clean session instance, and thus - callback